resin/logs.py

```diff
@@ -1,133 +1,188 @@
+    def subscribe(self, uuid, callback, error=None, count=None):
         """
-        This function allows subscribing to device logs.
-        There are fields (`m`, `t`, `s`, `c`) in the output which can be unclear. They stand for:
-            m - The log message itself.
-            t - The log timestamp.
-            s - Is this a system message?
-            c - The id of the service which produced this log (or null if the device does not support multiple containers).
+        Subscribe to device logs.
 
         Args:
             uuid (str): device uuid.
-            callback (function): this callback is called on receiving a message from the channel.
-            error (function): this callback is called on an error event.
-            For more details about callbacks in pubnub subscribe, visit here: https://www.pubnub.com/docs/python/api-reference#subscribe
-
-        Examples:
-            # Define callback and error.
-            >>> def callback(message, channel):
-            ...     print(message)
-            >>> def error(message):
-            ...     print('Error:'+ str(message))
-            >>> Logs.subscribe(uuid=uuid, callback=callback, error=error)
-
-        """
+            callback (function): this callback is called on receiving a message.
+            error (Optional[function]): this callback is called on an error event.
+            count (Optional[int]): number of historical messages to include.
 
-        _print_deprecation_warning()
-        channel = self.get_channel(uuid)
-        self.pubnub.subscribe(channels=channel, callback=callback, error=error)
+        Returns:
+            dict: a log entry will contain the following keys: `isStdErr, timestamp, message, isSystem, createdAt`.
 
-    @_init_pubnub
-    def history(self, uuid, callback, error):
         """
-        This function allows fetching historical device logs.
-
-        Args:
-            uuid (str): device uuid.
-            callback (function): this callback is called on receiving a message from the channel.
-            error (function): this callback is called on an error event.
-            For more details about callbacks in pubnub subscribe, visit here: https://www.pubnub.com/docs/python/api-reference#history
-
-        Examples:
-            # Define callback and error.
-            >>> def callback(message):
-            ...     print(message)
-            >>> def error(message):
-            ...     print('Error:'+ str(message))
-            Logs.history(uuid=uuid, callback=callback, error=error)
 
-        """
+        self.device.get(uuid)
+        if uuid is not None and uuid not in self.subscriptions:
+            self.subscriptions[uuid] = self.subscription_handler.add(uuid, callback, error, count)
 
-        _print_deprecation_warning()
-        channel = self.get_channel(uuid)
-        self.pubnub.history(channel=channel, callback=callback, error=error)
+        self.subscription_handler.run()
```

Comment:
This implementation still spawns one thread per connection. That's not how twisted should be used and ideally you shouldn't be dealing with threads at all.

Reply:
This call won't spawn a new thread per connection, it only kicks off the reactor if it's not running.